### ISSUE TITLE
@Scheduled with fixedDelay fails on startup

### ISSUE THREAD

**Issue by sddakoty** [spring-projects/spring-boot#2310]
Example project: https://drive.google.com/file/d/0B0qVi9D7R5MZeUYyeVlybkhhR3M/view?usp=sharing

Main Problem:
When scheduling a job with `@Scheduled(fixedDelay)`, it will be executed immediatelly. Since spring-boot.1.2.0.RELEASE this does not work anymore and fails with the following exception.

The issue can be temporary fixed by appending `@Scheduled(fixedDelay, initialDelay)`, but that should just be a workaround.

The example project demostrates that a simple functionless job is only found when using the initialDelay.
```
2015-01-08 15:06:40.284  INFO 12440 --- [           main] o.s.c.support.DefaultLifecycleProcessor  : Starting beans in phase 0
2015-01-08 15:06:40.284 ERROR 12440 --- [pool-1-thread-1] o.s.s.s.TaskUtils$LoggingErrorHandler    : Unexpected error occurred in scheduled task.

org.springframework.batch.core.launch.NoSuchJobException: No job configuration with the name [MyJobConfig] was registered
	at org.springframework.batch.core.configuration.support.MapJobRegistry.getJob(MapJobRegistry.java:66)
	at sample.actuator.log4j2.PeriodicJobStarter.runJob(PeriodicJobStarter.java:30)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:65)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:304)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:178)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:744)
```

**Comment by philwebb** [comment]
Did this work with Spring Boot 1.1?

**Comment by sddakoty** [comment]
As far as I remember it did, yes. I was surprised that this error occurs occasionally, since the upgrade to 1.2.0.

**Comment by wilkinsona** [comment]
Thanks for the sample that reproduces the problem; it's proven very useful.

In Spring Boot 1.2.x there is a race between the registration of the job and the scheduler calling `PeriodicJobStarter`. In Spring Boot 1.1.x this race condition either isn't there, or the registration of the job wins as the scheduling failure doesn't occur. Interestingly, the problem occurs with Boot 1.1.10 if the Spring Framework version's overridden to `4.1.4.RELEASE`, i.e. the change in behaviour seems to be due to a change in Spring Framework.

**Comment by wilkinsona** [comment]
Spring Batch performs it job registration in response to a `ContextRefreshedEvent`. In Spring 4.0, the scheduler struts calling scheduled tasks in response to a `ContextRefreshedEvent`. In Spring 4.1 the latter has changed due to https://github.com/spring-projects/spring-framework/commit/92c657e12d9a9e6da7ee22e7147e16ac7a11880b. It made `ScheduledAnnotationBeanPostProcessor` a `SmartInitializingSingleton`. In short this means that the scheduler now starts calling scheduled tasks before the `ContextRefreshedEvent` is sent. This means that `PeriodicJobStarter` is called before Batch has had a chance to register the job.

**Comment by wilkinsona** [comment]
I've opened [SPR-12641](https://jira.spring.io/browse/SPR-12641).

**Comment by wilkinsona** [comment]
The underlying issue in Spring Framework has been fixed and is available in 4.1.5 snapshots. It'll be fixed in Boot by #2413.
